nucypher/datastore/datastore.py

```diff
@@ -14,269 +14,222 @@
+
+        If the record is used outside the scope of the context manager, any
+        writes or reads will error.
+        """
+        with suppress(ValueError):
+            # If the ID can be converted to an int, we do it.
+            record_id = int(record_id)
+
+        with self.__db_env.begin(write=writeable) as datastore_tx:
+            record = record_type(datastore_tx, record_id, writeable=writeable)
+            try:
+                yield record
+            except (AttributeError, TypeError, DBWriteError) as tx_err:
+                # Handle `RecordNotFound` cases when `writeable` is `False`.
+                if not writeable and isinstance(tx_err, AttributeError):
+                    raise RecordNotFound(tx_err)
+                raise DatastoreTransactionError(f'An error was encountered during the transaction (no data was written): {tx_err}')
+            finally:
+                # Now we ensure that the record is not writeable
+                record.__dict__['_DatastoreRecord__writeable'] = False
+
+    @contextmanager
+    def query_by(self,
+              record_type: 'DatastoreRecord',
+              filter_func: Callable[[Union[Any, 'DatastoreRecord']], bool] = None,
+              filter_field: str = "",
+              writeable: bool = False,
+              ) -> List['DatastoreRecord']:
+        """
+        Performs a query on the datastore for the record by `record_type`.
+
+        An optional `filter_func` callable will take the decoded field
+        specified by the optional arg `filter_field` (see below) for the given
+        `record_type` iff the `filter_field` has been provided.
+        If no `filter_field` has been provided, then the `filter_func` will
+        receive a _readonly_ `DatastoreRecord`.
+
+        An optional `filter_field` can be provided as a `str` to perform a
+        query on a specific field for a `record_type`. This will cause the
+        `filter_func` to receive the decoded `filter_field` per the `record_type`.
+        Additionally, providing a `filter_field` will limit the query to
+        iterating over only the subset of records specific to that field.
+
+        If records can't be found, this method will raise `RecordNotFound`.
+        """
+        valid_records = set()
+        with self.__db_env.begin(write=writeable) as datastore_tx:
+            db_cursor = datastore_tx.cursor()
+
+            # Set the cursor to the closest key (if it exists) by the query params.
+            #
+            # By providing a `filter_field`, the query will immediately be
+            # limited to the subset of keys for the `filter_field`.
+            query_key = f'{record_type.__name__}:{filter_field}'.encode()
+            if not db_cursor.set_range(query_key):
+                # The cursor couldn't identify any records by the key
+                raise RecordNotFound(f"No records exist for the key from the specified query parameters: '{query_key}'")
```

Comment:
There are several places with this error text , maybe extract to constant/error?